src/Microsoft.Diagnostics.DebugServices.Implementation/ServiceContainer.cs

```diff
@@ -0,0 +1,184 @@
+            {
+                services = new List<ServiceFactory>();
+                _factories.Add(type, services);
+            }
+            services.Add(factory);
+        }
+
+        /// <summary>
+        /// Add a service instance. Multiple instances for the same type are not allowed.
+        /// </summary>
+        /// <param name="type">service type</param>
+        /// <param name="service">service instance (must derives from type)</param>
+        public void AddService(Type type, object service) => _instances.Add(type, service);
+
+        /// <summary>
+        /// Flushes the cached service instance for the specified type. Does not remove the service factory registered for the type.
+        /// </summary>
+        /// <param name="type">service type</param>
+        public void RemoveService(Type type) => _instances.Remove(type);
+
+        /// <summary>
+        /// Flushes all the cached instances of the services. Does not remove any of the service factories registered.
+        /// </summary>
+        public void FlushServices() => _instances.Clear();
+
+        /// <summary>
+        /// Dispose of the instantiated services.
+        /// </summary>
+        public void DisposeServices()
+        {
+            foreach (object service in _instances.Values)
+            {
+                if (service is IDisposable disposable)
+                {
+                    disposable.Dispose();
+                }
+            }
+            _instances.Clear();
+        }
+
+        /// <summary>
+        /// Get the cached/instantiated service instance if one exists. Don't call the factory or parent to create.
+        /// </summary>
+        /// <param name="type">service type</param>
+        /// <param name="service">service instance (can be null)</param>
+        /// <returns>if true, found service</returns>
+        public bool TryGetCachedService(Type type, out object service)
+        {
+            Debug.Assert(type != null);
+            if (type == typeof(IServiceProvider))
+            {
+                service = _serviceProvider;
+                return true;
+            }
+            if (type == typeof(IServiceContainer))
+            {
+                service = this;
+                return true;
+            }
+            return _instances.TryGetValue(type, out service);
+        }
+
+        /// <summary>
+        /// Returns the instance of the service or returns null if service doesn't exist
+        /// </summary>
+        /// <param name="type">service type</param>
+        /// <returns>service instance or null</returns>
+        public object GetService(Type type)
+        {
+            if (TryGetCachedService(type, out object service))
+            {
+                return service;
+            }
+            bool isEnumerable = type.IsGenericType && typeof(IEnumerable<>).IsAssignableFrom(type.GetGenericTypeDefinition());
```

Comment:
Shouldn't a check on the parent's go first? The problem is we depend on this implementation detail with the different `MemoryServices`

Reply:
We start with the leaf node service provider and chain up to the global one.

Reply:
I think if the container has a factory then that is the one that provides/instantiate the service.